### PR TITLE
[Feature] Add support for keyed services

### PR DESCRIPTION
**Pull request by martintmk**
# Pull Request

## The issue or feature being addressed

#1874 

## Details on the issue fix or feature implementation

This PR adds support for keyed services into Polly.

## Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

File: docs/advanced/dependency-injection.md [hunk 1]
```diff
@@ -100,6 +100,56 @@ await pipeline.ExecuteAsync(
...
+        // One is generic, the other is not.
+        _pipeline = pipeline;
+        _genericPipeline = genericPipeline;
+    }
+}
+```
+<!-- endSnippet -->
+
+> [!NOTE]
+> The resilience pipelines are registered in the DI container as transient services. This enables the resolution of multiple instances of `ResiliencePipeline` when [complex pipeline keys](#complex-pipeline-keys) are used. Resilience pipeline is retrieved and registered using `ResiliencePipelineProvider` that is responsible for lifetime management of resilience pipelines.
```
Comment (martincostello) [inline docs/advanced/dependency-injection.md:151]:
```suggestion
> The resilience pipelines are registered in the DI container as transient services. This enables the resolution of multiple instances of `ResiliencePipeline` when [complex pipeline keys](#complex-pipeline-keys) are used. The resilience pipeline is retrieved and registered using `ResiliencePipelineProvider` that is responsible for lifetime management of resilience pipelines.
```

File: docs/advanced/dependency-injection.md [hunk 2]
```diff
@@ -133,6 +183,9 @@ services
 ```
 <!-- endSnippet -->
 
+> [!NOTE]
+> The `AddResiliencePipelines` method does not support keyed services. To enable the resolution of a resilience pipeline using keyed services, you should use the `AddResiliencePipeline` extension, which adds a single resilience pipeline and registers it into the keyed services.
```
Comment (martincostello) [inline docs/advanced/dependency-injection.md:187]:
```suggestion
> The `AddResiliencePipelines` method does not support keyed services. To enable the resolution of a resilience pipeline using keyed services, you should use the `AddResiliencePipeline` extension method, which adds a single resilience pipeline and registers it into the keyed services.
```

**Review by martincostello — APPROVED** [review]

File: docs/advanced/dependency-injection.md (hunk 2 above)
Comment (martintmk) [reply docs/advanced/dependency-injection.md]: Done.

File: docs/advanced/dependency-injection.md (hunk 1 above)
Comment (martintmk) [reply docs/advanced/dependency-injection.md]: Done.

**Comment by pekspro** [comment]
I just want to say thank you for this. I did not expect this at all :-)

**Comment by hayohayo** [comment]
I did not investigate further, but when I update `Polly.Extensions` from 8.2.0 to 8.3.0 I start getting this exception:

```
System.InvalidOperationException: This service descriptor is keyed. Your service provider may not support keyed services.
   at Microsoft.Extensions.DependencyInjection.ServiceDescriptor.ThrowKeyedDescriptor()
   at Microsoft.Extensions.DependencyInjection.ServiceDescriptor.get_ImplementationInstance()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.Populate()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory..ctor(ICollection`1 descriptors)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider..ctor(ICollection`1 serviceDescriptors, ServiceProviderOptions options)
   at Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions.BuildServiceProvider(IServiceCollection services, ServiceProviderOptions options)
   at Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions.BuildServiceProvider(IServiceCollection services)
   ...
```

I think it might be caused by this PR. I should mention, this application is still using .NET 6. In .NET 8 I don't see this exception.

The way I use the resilience pipeline is something like:

```csharp
// On launch
services.AddResiliencePipeline<string, SomeResult>("someKey", (builder, context) => { /* ... */ });

// Inject into a singleton
class MySingleton
{
    public MySingleton(ResiliencePipelineProvider<string> resiliencePipelineProvider) { }
}
```

**Comment by CSharpFiasco** [comment]
> I did not investigate further, but when I update `Polly.Extensions` from 8.2.0 to 8.3.0 I start getting this exception:
> 
> ```
> System.InvalidOperationException: This service descriptor is keyed. Your service provider may not support keyed services.
>    at Microsoft.Extensions.DependencyInjection.ServiceDescriptor.ThrowKeyedDescriptor()
>    at Microsoft.Extensions.DependencyInjection.ServiceDescriptor.get_ImplementationInstance()
>    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.Populate()
>    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory..ctor(ICollection`1 descriptors)
>    at Microsoft.Extensions.DependencyInjection.ServiceProvider..ctor(ICollection`1 serviceDescriptors, ServiceProviderOptions options)
>    at Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions.BuildServiceProvider(IServiceCollection services, ServiceProviderOptions options)
>    at Microsoft.Extensions.DependencyInjection.ServiceCollectionContainerBuilderExtensions.BuildServiceProvider(IServiceCollection services)
>    ...
> ```
> 
> I think it might be caused by this PR. I should mention, this application is still using .NET 6. In .NET 8 I don't see this exception.
> 
> The way I use the resilience pipeline is something like:
> 
> ```cs
> // On launch
> services.AddResiliencePipeline<string, SomeResult>("someKey", (builder, context) => { /* ... */ });
> 
> // Inject into a singleton
> class MySingleton
> {
>     public MySingleton(ResiliencePipelineProvider<string> resiliencePipelineProvider) { }
> }
> ```

I believe this is related https://github.com/dotnet/runtime/issues/95789

**Pull request merged**